File: generate-parentheses.py
from typing import List


class Solution:
    def __init__(self):
        self.result = []

    def generateParenthesis(self, n: int) -> List[str]:
        self.result = []
        self.backtrack("(", 1, 0, n)
        return self.result

    def backtrack(self, curr: str, count_open: int, count_closed: int, n: int):
        if count_open == n and count_closed == n:
            self.result.append(curr)
            return

        if count_open < n:
            self.backtrack(curr + "(", count_open + 1, count_closed, n)

        if count_closed < count_open:
            self.backtrack(curr + ")", count_open, count_closed + 1, n)

File: solutions/generate-parentheses.java
import java.util.List;
import java.util.ArrayList;


class Solution {
    private List<String> result = new ArrayList<>();

    public List<String> generateParenthesis(int n) {
        this.backtrack("(", 1, 0, n);
        return result;
    }

    private void backtrack(String curr, int countOpen, int countClosed, int n) {
        if (countOpen == n && countClosed == n) {
            this.result.add(curr);
            return;
        }

        if (countOpen < n) {
            this.backtrack(curr + "(", countOpen + 1, countClosed, n);
        }

        if (countClosed < countOpen) {
            this.backtrack(curr + ")", countOpen, countClosed + 1, n);
        }
    }
}
